File: my-folder/0543-diameter-of-binary-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
    diameter := 0

    dfs(root, &diameter)
    return diameter
}

func dfs(node *TreeNode, diameter *int) int {
    if node == nil {
        return 0
    }
    left := dfs(node.Left, diameter)
    right := dfs(node.Right, diameter)
    *diameter = max(*diameter, left + right)

    return 1 + max(left, right)
}

File: my-folder/0543-diameter-of-binary-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    let diameter = 0

    let dfs = function(root: TreeNode | null) {
        if (root == null) {
            return 0
        }

        let left = dfs(root.left)
        let right = dfs(root.right)

        diameter = Math.max(diameter, left+right)

        return 1 + Math.max(left, right)
    }

    dfs(root)

    return diameter
};


